lib/LoadProgress.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * LoadProgress.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "StdInc.h"
+#include "LoadProgress.h"
+
+using namespace Load;
+
+Progress::Progress(): _progress(std::numeric_limits<Type>::min())
+{
+	steps(100);
+}
+
+Type Progress::get() const
+{
+	return _progress;
+}
+
+void Progress::set(Type p)
+{
+	_progress = p;
+}
+
+bool Progress::finished() const
+{
+	return _progress == std::numeric_limits<Type>::max();
+}
+
+void Progress::reset(int s)
+{
+	_progress = std::numeric_limits<Type>::min();
+	steps(s);
+}
+
+void Progress::finish()
+{
+	_progress = std::numeric_limits<Type>::max();
+}
+
+void Progress::steps(int s)
+{
+	stepsTill(s, std::numeric_limits<Type>::max());
+}
+
+void Progress::stepsTill(int s, Type p)
+{
+	if(p < _progress)
+		_step = 0;
+	else
+		_step = (p - _progress) / s;
+}
+
+void Progress::step(int count)
+{
+	for(int i = 0; i < count; ++i)
+	{
+		if(_progress + _step < _progress)
```

Comment:
byte + byte is still byte? No automatic cast to int in c++? Anyway I suggest to not relay on overflow behavior. Also  we can store progress as double 0 - 100 and double allows fractions so step can be less than 1,

Reply:
And we will be able to check if(_progress + _step * count > 100) finish();

Reply:
same way we could use `uint32_t`, treat 100% e.g. as 100*1000 and avoid all floating-point issues

Reply:
Come on guys, just count to 100 or 1000, who cares?

Reply:
Just a suggestion. 100% is quite natural and performance is not critical here but fractions of % might be needed

Reply:
What if _step becomes 0 due to numeric /? No progress will be until next stepsTill

Reply:
> byte + byte is still byte? No automatic cast to int in c++?
Nope, C++ casts result to the type of left operand

I don't want to make any of constants like 100 because you may want to have intermediate values. It can be solved with double but as @kambala-decapitator mentionned we might have a problem with float-point computations and compiler-depended behavior

Reply:
but there shouldn't be any issues with byte too, anyway if we have neccessity to change type in the future, it can be done with single line